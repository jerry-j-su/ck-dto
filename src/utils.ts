export function simpleUID(letter: number) {
    const letterCount =  (!Number.isInteger(letter) || letter < 6 || letter > 11) ? 6 : letter
    return Math.random().toString(32).substring(2, 2 + letterCount);
}

/**
 * A simple helper to transform a function to tailing debounced version, avoid cost of high-frequency input
 * @param {Function} func 
 * @param {number} wait - waiting time (in ms) to delay the execution 
 */
export function tailingDebounce(func: Function, wait: number) {
    let timerId: number
    let latestArgs: any[]
    let latestThis: any
    
    function startTimer() {
        window.clearTimeout(timerId)
        timerId = window.setTimeout(() => {
            func.apply(latestThis, latestArgs)
        }, wait)
    }

    function debounced(...args: any[]) {
        latestArgs = args
        // @ts-ignore
        latestThis = this
        startTimer();
    }

    return debounced
}