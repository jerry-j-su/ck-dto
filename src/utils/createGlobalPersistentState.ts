import { useState, useEffect, useLayoutEffect } from 'react';

import { IGlobalStateInitiator, IGlobalStateSetter } from '../types'

function resolveNextState<S>(nextState: IGlobalStateSetter<S>): S {
    return typeof nextState === 'function' ? (nextState as Function)() : nextState
}

export function createGlobalPersistentState<S = any>(
    initialState: IGlobalStateInitiator<S>
): () => [S, (state: ((IGlobalStateSetter<S>))) => void];
export function createGlobalPersistentState<S = undefined>(): () => [
    S,
    (state: IGlobalStateSetter<S>) => void
];

/**
 * The useGlobalState hook factory
 * @param initialState
 * @returns
 */
export function createGlobalPersistentState<S>(initialState?: S) {
    /* Closure to persist a global store for the given state */
    const store: {
        state: S;
        setState: (state: IGlobalStateSetter<S>) => void;
        setters: any[];
    } = {
        state: initialState instanceof Function ? initialState() : initialState,
        // broadcast the state change to all components accessing this global state, so that they can be updated
        setState(nextState: IGlobalStateSetter<S>) {
            store.state = resolveNextState(nextState);
            store.setters.forEach((setter) => setter(store.state));
        },
        // setter list, any component connects to this global state must register its own state setter here
        setters: [],
    };

    /*
      the hook for Component to access the global state
      It is private React.useState that initialized and sync with the global state value
      usage is similar to React.useState
    */
    return () => {
        const [globalState, stateSetter] = useState<S | undefined>(store.state);

        // clean up registered setter upon component unmount
        useEffect(() => () => {
            store.setters = store.setters.filter((setter) => setter !== stateSetter);
        }, []);

        // register the setter to the setter list so that change from other component can trigger the state setter of this component
        useLayoutEffect(() => { // eslint-disable-line react-hooks/exhaustive-deps
            if (!store.setters.includes(stateSetter)) {
                store.setters.push(stateSetter);
            }
        });

        return [globalState, store.setState];
    };
}

export default createGlobalPersistentState;
