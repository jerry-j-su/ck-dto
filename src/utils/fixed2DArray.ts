const DEFAULT_BLOCK_SIZE = 1000

export interface IFixed2DArray<I> {
    get: (index: number) => I | undefined,
    length: number,
    add: (item: I) => number,
    set: (index: number, item: I) => IFixed2DArray<I>,
    toArray: () => I[],
    slice: (start?: number, end?: number) => I[],
}

export default function fixed2DArray<I> ({ blockSize = DEFAULT_BLOCK_SIZE }: { blockSize?: number } = {}): IFixed2DArray<I> {
    if (!Number.isInteger(blockSize) || blockSize <= 0) {
        throw new Error('Invalid block size')
    }

    const _blockSize = blockSize        // the fixed block size (width of each row)
    const _table: I[][] = []            // internal table for data storage
    const _pos = { x: 0, y: 0 }         // pointer
    let _length = 0                     // item count

    /**
     * Move position forward, warp it to the beginning of the next new block upon reaching the end of previous one
     */
    const forward = () => {
        if (_pos.x < _blockSize - 1) _pos.x ++
        else {
            _pos.x = 0
            _pos.y ++
        }
        _length ++
    }

    return {
        /**
         * Return the indexed item, index is the flatten position
         * @param {number} index
         */
        get(index: number) {
            if (index < 0 || index >= this.length) {
                return undefined
            }
            const x = index % _blockSize
            const y = Math.floor(index / _blockSize)
            return _table[y][x]
        },
        /**
         * Return the size of the 2D array
         */
        get length() {
            return _length
        },
        /**
         * Add an item in the next available position
         * @param {I} item
         */
        add(item: I) {
            if (!_table[_pos.y]) {
                _table[_pos.y] = new Array(_blockSize)
            }
            _table[_pos.y][_pos.x] = item
            forward()
            return this.length
        },
        /**
         * Modify an existing item in the 2D array
         * @param {number} index
         * @param {I} item
         */
        set(this: IFixed2DArray<I>, index: number, item: I): IFixed2DArray<I> {
            if (index < 0 || index >= this.length) {
                throw new Error('out-of-range error')
            }
            const x = index % _blockSize
            const y = Math.floor(index / _blockSize)
            _table[y][x] = item
            return this
        },
        /**
         * Return a primitive array containing a portion of the items in the table. The portion is specified by the start and end index
         * @param {number} start - the flattened index of start position
         * @param {number} end - the flattened index of end position
         */
        slice(this: IFixed2DArray<I>, start: number = 0, end: number = this.length) {
            const actualStart = start >= 0
                ? Math.min(start, this.length - 1)
                : Math.max(-this.length, start) + this.length
            const actualEnd = end >= 0
                ? Math.min(end, this.length)
                : Math.max(-this.length, end) + this.length

            if (actualStart >= actualEnd) return []

            const startX = actualStart % _blockSize
            const startY = Math.floor(actualStart / _blockSize)
            const endX = actualEnd % _blockSize
            const endY = Math.floor(actualEnd / _blockSize)

            const moreThanOneBlock = endY > startY
            return _table
                .slice(startY, endY + 1)
                .reduce((acc, block, index, subset) => {
                    switch (index) {
                        case 0:
                            return acc.concat(block.slice(startX, moreThanOneBlock ? _blockSize : endY))
                        case subset.length - 1:
                            return acc.concat(block.slice(0, endX || _blockSize))
                        default:
                            return acc.concat(block)
                    }
                }, [])
        },
        /**
         * Returns the flatten primitive array for UI to consume
         * @returns
         */
        toArray() {
            return _table.reduce((acc, block, blockIndex) => {
                return (blockIndex < _pos.y) ? acc.concat(block) : acc.concat(block.filter(Boolean))
            }, [])
        },

        // map() {

        // },

        // reverse() {

        // },

        // debug
        // @ts-ignore
        // p_table() {
        //     console.log(_table)
        // },
    }
}
