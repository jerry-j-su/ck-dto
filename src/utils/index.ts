export * from './misc'
export { default as createGlobalPersistentState } from './createGlobalPersistentState'
export { default as fixed2DArray } from './fixed2DArray'
export * from './fixed2DArray'
export { default as reverseIndexMap } from './reverseIndexMap'
