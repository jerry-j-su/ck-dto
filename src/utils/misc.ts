export function simpleUID(letter: number) {
    const letterCount = (!Number.isInteger(letter) || letter < 6 || letter > 11) ? 6 : letter
    return Math.random().toString(32).substring(2, 2 + letterCount)
}

/**
 * Merge multiple class names into a consumable className property
 * @param {string[]} classNames
 */
export function mergeClassNames(classNames: (string | undefined)[]) {
    if (!Array.isArray(classNames)) return ''
    return classNames.filter(Boolean).join(' ')
}

type DebounceOptions = {
    leading?: boolean
    trailing?: boolean
    maxWait?: number
}
type DebouncedFunction<R> = (...args: any[]) => R
type ThrottleOptions = Omit<DebounceOptions, 'maxWait'>
type ThrottledFunction<R> = (...args: any[]) => R
/**
 * Transform a function to its trailing debounced variant, avoid cost of high-frequency input
 * @param {Function} func - function to be debounced
 * @param {number} wait - waiting time (in ms) to delay the execution
 * @param {boolean} [options.leading = false] - execute at leading edge
 * @param {boolean} [options.trailing = true] - execute at trailing edge
 * @param {boolean} [options.maxWait] - the maximum interval between each debounced executions
 */
export function debounce<R>(
    func: (...args: any[]) => R,
    wait: number,
    { leading = false, trailing = true, maxWait = 0 }: DebounceOptions = {},
): DebouncedFunction<R> {
    let timerId: number
    let latestArgs: any
    let latestThis: Object
    let isWaiting = false
    let invokeNow: (...args: any[]) => R
    let lastExecTime: number
    let lastResult: R
    const maxWaitIsSet = Number.isInteger(maxWait) && maxWait > 0

    function shouldInvokeNow() {
        return !isWaiting && leading
    }

    function resetCycleForLongPause() {
        // check if previous wait is long enough to reset the cycle
        if (maxWaitIsSet && lastExecTime && Date.now() - lastExecTime > maxWait) {
            isWaiting = false
            lastExecTime = Date.now()
        }
    }

    function startTimer() {
        // clear current timer
        window.clearTimeout(timerId)

        // calculate waiting time
        let timeSinceLastExec = lastExecTime ? Date.now() - lastExecTime : 0
        let remainingWait = maxWaitIsSet
            ? Math.min(wait, maxWait - timeSinceLastExec)
            : wait

        timerId = window.setTimeout(() => {
            // trailing edge
            if (trailing) {
                lastResult = invokeNow.apply(latestThis, latestArgs)
                lastExecTime = Date.now()
            }

            isWaiting = (leading && trailing && maxWaitIsSet)
                ? true
                : false
        }, remainingWait)
    }

    // this is the debounced method to invoke original execution at a timely manner
    function debounced(this: Object, ...args: any[]): R {
        resetCycleForLongPause()
        // leading edge
        if (shouldInvokeNow()) {
            lastResult = func.apply<Object, any[], R>(this, args)
            lastExecTime = Date.now()
        }
        invokeNow = func
        latestThis = this
        latestArgs = args
        isWaiting = true
        startTimer()
        return lastResult
    }

    return debounced
}

/**
 * Transform a function to its throttled variant, cap the frequency of execution
 * @param {Function} func - function to be debounced
 * @param {number} wait - waiting time (in ms) to delay the execution
 * @param {boolean} [options.leading = true] - execute at leading edge
 * @param {boolean} [options.trailing = true] - execute at trailing edge
 */
export function throttle<R>(
    func: (...args: any[]) => R,
    wait: number,
    { leading = true, trailing = true }: ThrottleOptions = {},
): ThrottledFunction<R> {
    return debounce(func, wait, { leading, trailing, maxWait: wait })
}

export const observeDOM = (() => {
    const MutationObserver = window.MutationObserver

    return function (ele: HTMLElement, callback: Function) {
        if (!(ele instanceof HTMLElement) || ele.nodeType !== 1) return

        if (MutationObserver) {
            const mutationObserver = new MutationObserver((mutationList: MutationRecord[]) => {
                callback()
            })

            // have the observer observe for DOM changes in children
            mutationObserver.observe(ele, { childList: true, subtree: true })
            return mutationObserver
        }

        // Older browser that does not support MutationObserver
        else if (ele.addEventListener) {
            ele.addEventListener('DOMNodeInserted', callback as EventListener, false)
            ele.addEventListener('DOMNodeRemoved', callback as EventListener, false)
        }
    }
})()
