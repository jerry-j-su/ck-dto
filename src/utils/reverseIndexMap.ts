/**
 * Inverse index hash table for quick search
 */
function reverseIndexMap<KeyType, I> () {
    const _map = new Map<KeyType, I[]>()

    return {
        clear() {
            const keys = _map.keys()
            let isDone = false
            do {
                const { value: key, done } = keys.next()
                _map.delete(key)
                isDone = !!done
            }
            while (!isDone)
        },

        get: _map.get.bind(_map),

        put(key: KeyType, index: I) {
            const bucket = _map.get(key)
            if (Array.isArray(bucket)) bucket.push(index)
            else _map.set(key, [index])
        },

        keys: _map.keys.bind(_map),

        /* methods that are not needed */

        // size(key: KeyType) {
        //     if (key) return (_map.get(key) || []).length
        //     let length = 0
        //     for(const bucket of _map.values()) {
        //         length += (bucket || []).length
        //     }
        //     return length;
        // },

        // delete(key: KeyType) {},

        // has(key: KeyType) {
        //     return Array.isArray(_map[key] && _map[key].length > 0)
        // }

        // entries: _map.entries.bind(_map),
    }
}

export default reverseIndexMap;
