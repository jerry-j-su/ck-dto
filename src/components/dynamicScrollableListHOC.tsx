/**
 * Dynamic Scrollable list for long list:
 * - Increased Render Performance:
 *   DOM nodes count is limited on certain scroll range. Wrapped Component's DOM creation will be held upon scrolling to certain position
 * - Enables specific scroll handle/behavior for the Wrapped Component DOM
 * - handler is debounced/throttled
 */
import React, { useEffect, useRef, useState, useCallback } from 'react'

import { mergeClassNames, observeDOM, throttle } from '../utils/misc'

export type ScrollableOptions = {
    interval?: number,
}
enum ToBottom { InRange, Near, Far }
export type ViewPosition = {
    atTop: boolean,
    toBottom: ToBottom,
}

// property to pass to its wrapped component, telling it to hold/continue render new content DOM nodes
export type ScrollableContainerExtraProps = {
    holdAppendChild?: boolean
}

/* Default settings */
const DEFAULT_STYLE = {
    wrapper: 'scrollable-wrapper',
    tips: {
        top: 'scrollable-tip-top',
        bottom: 'scrollable-tip-bottom',
    },
}
const DEFAULT_OPTIONS: ScrollableOptions = {
    interval: 50,
}
const DEFAULT_VIEW_POSITION = { atTop: true, toBottom: ToBottom.InRange }
Object.freeze(DEFAULT_STYLE)
Object.freeze(DEFAULT_OPTIONS)
Object.freeze(DEFAULT_VIEW_POSITION)

/**
 * Calculate the position of the viewable part relative to the entire long content DOM
 * @param {HTMLElement} containerDOM - the limited size viewport container that scrolls
 * @param {HTMLElement} contentDOM - the long DOM that holds large amount of content
 * @returns
 */
function calculateViewPosition (containerDOM: HTMLElement, contentDOM: HTMLElement): ViewPosition {
    if (!(containerDOM instanceof HTMLElement) || !(contentDOM instanceof HTMLElement)) return DEFAULT_VIEW_POSITION

    // console.log('calculating view position')
    const { scrollTop } = containerDOM
    const { height: containerH } = containerDOM.getBoundingClientRect()
    const { height: contentH } = contentDOM.getBoundingClientRect()
    // make sure the contain has no padding nor margin
    const remainingH = Math.max(Math.floor(contentH - (scrollTop + containerH)), 0)
    const atTop = scrollTop <= 150

    if (remainingH <= 159) {
        // console.log(`at top: ${atTop}, ToBottom: in range`)
        return { atTop, toBottom: ToBottom.InRange }
    }
    if (remainingH <= 2 * containerH) {
        // console.log(`at top: ${atTop}, ToBottom: near`)
        return { atTop, toBottom: ToBottom.Near }
    }
    // console.log(`at top: ${atTop}, ToBottom: far`)
    return { atTop, toBottom: ToBottom.Far }
}


export type DynamicScrollableListWrapperProps = React.PropsWithoutRef<any> & {
    classes?: {
        wrapper?: string,
        tips?: {
            top?: string,
            bottom?: string,
        }
    },
}
/**
 * Dynamic Scrollable list for long list
 * @param {React.FC} Container - Wrapped Container Component which contains a possibly long-list node
 * @param {ScrollableOptions} [options]
 * @returns
 */
export default function dynamicScrollableListHOC(Container: any, options?: ScrollableOptions) {
    // construct a throttled version of the cost-heavy view calculate method
    const interval = (options?.interval || DEFAULT_OPTIONS.interval) as number
    const throttledCalculateViewPosition = throttle<ViewPosition>(calculateViewPosition, interval)

    return (props: DynamicScrollableListWrapperProps) => {
        const containerRef = useRef<HTMLElement>(null)
        const containerDOM = containerRef?.current
        const contentDOM = containerDOM?.children[0] as HTMLElement

        const [viewPosition, setViewPosition] = useState<ViewPosition>(DEFAULT_VIEW_POSITION)

        const {
            classes: { wrapper, tips } = {},
            ...passThroughProps
        } = props

        const wrapperClass = mergeClassNames([wrapper, DEFAULT_STYLE.wrapper])
        const topTipClass = mergeClassNames([tips?.top, DEFAULT_STYLE.tips.top])
        const bottomTipClass = mergeClassNames([tips?.bottom, DEFAULT_STYLE.tips.bottom])

        /**
         * Handler upon scroll or list modification, throttled
         * Calculate the position of the viewable part relative to the entire long content DOM
         * Result will be passed to its wrapped component to let it hold/continue adding more content nodes
         */
        // eslint-disable-next-line react-hooks/exhaustive-deps
        const onScrollOrContentChange = useCallback(() => {
            setViewPosition(throttledCalculateViewPosition(containerDOM, contentDOM) || DEFAULT_VIEW_POSITION)
        }, [setViewPosition, containerDOM, contentDOM])

        useEffect(() => {
            let observer: MutationObserver | undefined
            if (containerDOM instanceof HTMLElement) {
                containerDOM.addEventListener('scroll', onScrollOrContentChange)
                observer = observeDOM(containerDOM, onScrollOrContentChange)
            }
            return () => {
                containerDOM?.removeEventListener('scroll', onScrollOrContentChange)
                observer?.disconnect && observer.disconnect()
            }
        }, [containerDOM, contentDOM, onScrollOrContentChange])

        return (
            <div className={wrapperClass} >
                {!viewPosition.atTop && <div className={topTipClass}>
                    <i className="arrow-up">&gt;</i>
                    <br />
                    Latest orders
                </div>}
                <Container {...passThroughProps} ref={containerRef} holdAppendChild={viewPosition.toBottom === ToBottom.Far} />
                {viewPosition.toBottom !== ToBottom.InRange && <div className={bottomTipClass}>
                    More orders
                    <br />
                    <i className="arrow-down">&lt;</i>
                </div>}
            </div>
        )
    }
}
