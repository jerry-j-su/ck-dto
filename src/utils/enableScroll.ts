/**
 * Enables specific scroll behavior for given container DOM
 */
import { useEffect } from 'react'

import { mergeClassNames, observeDOM, tailingDebounce } from './misc'

enum ScrollType {
    auto = 'auto',
    hidden = 'hidden',
    visible = 'visible',
}
export type ScrollableOptions = {
    x?: ScrollType,
    y?: ScrollType,
    interval?: number,
    style?: {
        container?: string,
        trigger?: {
            top?: string,
            bottom?: string,
            left?: string,
            right?: string,
        }
    }
}
export type useScrollableReturns = {
    triggerClassNames: { top: string, bottom: string, left: string, right: string }
}

const DEFAULT_STYLE = {
    container: 'scrollable-container',
    trigger: {
        top: 'scrollable-top',
        bottom: 'scrollable-bottom',
        left: 'scrollable-left',
        right: 'scrollable-right',
    },
}
const DEFAULT_OPTIONS: ScrollableOptions = {
    x: ScrollType.auto,
    y: ScrollType.auto,
    interval: 50,
    style: DEFAULT_STYLE,
}
Object.freeze(DEFAULT_OPTIONS)

function onScroll(event: Event) {
    const { currentTarget } = event

    console.log(event);
}

export default function enableScroll(el: HTMLElement | null, givenOptions?: ScrollableOptions) {


    const options = {
        ...givenOptions,
        style: {
            container: mergeClassNames([givenOptions?.style?.container, DEFAULT_STYLE.container]),
            trigger: {
                top: mergeClassNames([givenOptions?.style?.trigger?.top, DEFAULT_STYLE.trigger.top]),
                bottom: mergeClassNames([givenOptions?.style?.trigger?.bottom, DEFAULT_STYLE.trigger.bottom]),
                left: mergeClassNames([givenOptions?.style?.trigger?.left, DEFAULT_STYLE.trigger.left]),
                right: mergeClassNames([givenOptions?.style?.trigger?.right, DEFAULT_STYLE.trigger.right]),
            },
        }
    }

    if (el instanceof HTMLElement) {
        el.addEventListener('scroll', onScroll)
        observeDOM(el, tailingDebounce(() => {
            console.log('contain added')
        }, 200))
    }

    return {
        triggerClassNames: {
            ...options.style.trigger
        },
        handlerFunc: onScroll,
    }
}
