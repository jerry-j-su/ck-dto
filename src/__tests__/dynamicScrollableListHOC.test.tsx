import React, { Children } from 'react'
import { render } from '@testing-library/react'
import { readFileSync } from 'fs';

import dynamicScrollableListHOC, { ToBottom, calculateViewPosition } from '../components/dynamicScrollableListHOC'

// module mocks
const sideEffectList: Function[] = []
const mockAddEventListener = jest.fn()
jest.mock('react', () => {
    const originalModule = jest.requireActual('react')
    return {
        __esModule: true,
        ...originalModule,
        useRef: () => ({
            current: {
                addEventListener: mockAddEventListener,
                children: [
                    {
                        getBoundingClientRect: () => ({ height: 4098 }),
                    }
                ]
            }
        }),
        useState: (initialValue: any) => [initialValue, () => { }],
        useCallback: () => () => { },
        useEffect: (cb: Function) => sideEffectList.push(cb),
    }
})
jest.mock('../utils/misc', () => ({
    __esModule: true,
    ...jest.requireActual('../utils/misc'),
    observeDOM: jest.fn()
}))
import { observeDOM as mockObserveDOM } from '../utils/misc'

const DummyComponent = () => (<div></div>)
const contentDOM = document.createElement('div')
const containerDOM = document.createElement('div')

describe('dynamicScrollableListHOC test', () => {
    beforeEach(() => {
        sideEffectList.splice(0) // clear side effect before each case
    })

    test('Basic wrapper render', () => {
        const nodeTree = dynamicScrollableListHOC(DummyComponent)({})
        expect(nodeTree).toMatchSnapshot();
    })

    test('Should begin monitoring content DOM mutation and scroll', () => {
        dynamicScrollableListHOC(DummyComponent)({})
        mockAddEventListener.mockClear()
        // @ts-ignore
        mockObserveDOM.mockClear()
        const sideEffect = sideEffectList.pop()
        if (sideEffect) sideEffect()
        expect(mockAddEventListener).toHaveBeenCalled()
        expect(mockObserveDOM).toHaveBeenCalled()

    })

    test('calculateViewPosition should detect when content bottom is in view area', () => {
        const contentDOMProxy = new Proxy(contentDOM, {
            get(target, prop, receiver) {
                if (prop === 'getBoundingClientRect') {
                    return () => ({ height: 900 })
                }
                // @ts-ignore
                return Reflect.get(...arguments);
            },
        });
        const containerDOMProxy = new Proxy(containerDOM, {
            get(target, prop, receiver) {
                switch (prop) {
                    case 'scrollTop':
                        return 0
                    case 'getBoundingClientRect':
                        return () => ({ height: 960 })
                    default:
                        // @ts-ignore
                        return Reflect.get(...arguments);
                }

            },
        });
        const viewPosition = calculateViewPosition(containerDOMProxy, contentDOMProxy)
        expect(viewPosition).toEqual({ atTop: true, toBottom: ToBottom.InRange })
    })

    test('calculateViewPosition should detect when view area is near content bottom', () => {
        const contentDOMProxy = new Proxy(contentDOM, {
            get(target, prop, receiver) {
                if (prop === 'getBoundingClientRect') {
                    return () => ({ height: 1440 })
                }
                // @ts-ignore
                return Reflect.get(...arguments);
            },
        });
        const containerDOMProxy = new Proxy(containerDOM, {
            get(target, prop, receiver) {
                switch (prop) {
                    case 'scrollTop':
                        return 200
                    case 'getBoundingClientRect':
                        return () => ({ height: 960 })
                    default:
                        // @ts-ignore
                        return Reflect.get(...arguments);
                }

            },
        });
        const viewPosition = calculateViewPosition(containerDOMProxy, contentDOMProxy)
        expect(viewPosition).toEqual({ atTop: false, toBottom: ToBottom.Near })
    })

    test('calculateViewPosition should detect when view area is far away from content bottom', () => {
        const contentDOMProxy = new Proxy(contentDOM, {
            get(target, prop, receiver) {
                if (prop === 'getBoundingClientRect') {
                    return () => ({ height: 4096 })
                }
                // @ts-ignore
                return Reflect.get(...arguments);
            },
        });
        const containerDOMProxy = new Proxy(containerDOM, {
            get(target, prop, receiver) {
                switch (prop) {
                    case 'scrollTop':
                        return 200
                    case 'getBoundingClientRect':
                        return () => ({ height: 960 })
                    default:
                        // @ts-ignore
                        return Reflect.get(...arguments);
                }

            },
        });
        const viewPosition = calculateViewPosition(containerDOMProxy, contentDOMProxy)
        expect(viewPosition).toEqual({ atTop: false, toBottom: ToBottom.Far })
    })
})
