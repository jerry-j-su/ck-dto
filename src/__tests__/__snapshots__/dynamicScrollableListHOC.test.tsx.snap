// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamicScrollableListHOC test Basic wrapper render 1`] = `
<div
  className="scrollable-wrapper"
>
  <DummyComponent
    holdAppendChild={false}
  />
</div>
`;
